clouddriver-kubernetes-v2/src/test/java/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesV2CredentialsTest.java

```diff
@@ -0,0 +1,376 @@
+                    NAMESPACE))
+        .isInstanceOf(KubectlException.class);
+    ImmutableList<Timer> timers = registry.timers().collect(toImmutableList());
+    assertThat(timers).hasSize(1);
+
+    Timer timer = timers.get(0);
+    assertThat(timer.id().name()).isEqualTo("kubernetes.api");
+
+    assertThat(timer.id().tags())
+        .containsExactlyInAnyOrder(
+            Tag.of("account", ACCOUNT_NAME),
+            Tag.of("action", "list"),
+            Tag.of("kinds", "deployment,replicaSet"),
+            Tag.of("namespace", NAMESPACE),
+            Tag.of("success", "false"),
+            Tag.of("reason", "CustomException: Kubernetes error"));
+  }
+
+  @Test
+  void timeRecordedForListThrowingOtherException() {
+    KubectlJobExecutor jobExecutor = mock(KubectlJobExecutor.class);
+
+    ManualClock clock = new ManualClock();
+    Registry registry = new DefaultRegistry(clock);
+    KubernetesV2Credentials credentials = getCredentials(registry, jobExecutor);
+
+    clock.setMonotonicTime(1000);
+    when(jobExecutor.list(eq(credentials), any(), any(), any()))
+        .then(
+            call -> {
+              clock.setMonotonicTime(1500);
+              throw new CustomException("Kubernetes errror");
+            });
+    assertThatThrownBy(
+            () ->
+                credentials.list(
+                    ImmutableList.of(KubernetesKind.DEPLOYMENT, KubernetesKind.REPLICA_SET),
+                    NAMESPACE))
+        .isInstanceOf(KubectlException.class);
+
+    ImmutableList<Timer> timers = registry.timers().collect(toImmutableList());
+    assertThat(timers).hasSize(1);
+
+    Timer timer = timers.get(0);
+    assertThat(timer.id().name()).isEqualTo("kubernetes.api");
+    assertThat(timer.totalTime()).isEqualTo(500);
+  }
+
+  @Test
+  void propagatedExceptionForListThrowingOtherException() {
+    KubectlJobExecutor jobExecutor = mock(KubectlJobExecutor.class);
+    Registry registry = new DefaultRegistry();
+    KubernetesV2Credentials credentials = getCredentials(registry, jobExecutor);
+
+    Exception cause = new CustomException("Kubernetes error");
+    when(jobExecutor.list(eq(credentials), any(), any(), any())).thenThrow(cause);
+
+    // Assert that a The source exception is the cause is passed through without modification
+    assertThatThrownBy(
+            () ->
+                credentials.list(
+                    ImmutableList.of(KubernetesKind.DEPLOYMENT, KubernetesKind.REPLICA_SET),
+                    NAMESPACE))
+        .hasMessageContaining("Failure running list")
+        .hasCause(cause);
+  }
+
+  // This is an error type that will only ever be thrown by stubs in this test; that way we can
+  // assert that it is thrown and be sure that we aren't accidentally passing due to an unrelated
+  // exception.
+  private static class CustomException extends RuntimeException {
+    CustomException(String message) {
+      super(message);
+    }
+  }
+}
```

Comment:
This is probably the most thorough set of tests I have ever encountered in Spinnaker ⛵️ 💯